File: 0229-majority-element-ii/0229-majority-element-ii.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        int limit = nums.length / 3;
        List<Integer> res = new ArrayList<>();
        HashMap<Integer, Integer> freq = new HashMap<>();
        for (int i : nums) {
            freq.put(i, freq.getOrDefault(i, 0) + 1);

            if(freq.size() <= 2) continue;

            HashMap<Integer, Integer> dummy = new HashMap<>();
            for(int key : freq.keySet()){
                if(freq.get(key) > 1){
                    dummy.put(key,freq.get(key)-1);
                }
            }
            freq = dummy;
        }

        for(int key : freq.keySet()){
            if(count(nums,key) > limit){
                res.add(key);
            }
        }

        return res;
    }

    private int count(int[] nums, int k) {
        int count = 0;
        for (int num : nums) {
            if (num == k) {
                count++;
            }
        }
        return count;
    }
}

File: 0229-majority-element-ii/0229-majority-element-ii.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        limit = len(nums) // 3
        freq = {}

        for num in nums:
            freq[num] = 1 + freq.get(num, 0)

            if len(freq) <= 2:
                continue

            temp = {}
            for key in freq:
                if freq[key] > 1:
                    temp[key] = freq[key] - 1
            freq = temp

        result = []

        for key in freq:
            count = 0
            for num in nums:
                if num == key:
                    count += 1

            if count > limit:
                result.append(key)

        return result
